File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii-Solution1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    const result = [];
    if(root === null) return result;

    let currLevel = [root];

    while(currLevel.length > 0){
        const newCurrLevel = [];
        const currLevelVal = [];
        for(let node of currLevel){
            currLevelVal.push(node.val);
            node.left !== null && newCurrLevel.push(node.left);
            node.right !== null && newCurrLevel.push(node.right);
        }
        currLevel = newCurrLevel;
        result.unshift(currLevelVal);
    }

    return result;
};

File: 0107-binary-tree-level-order-traversal-ii/0107-binary-tree-level-order-traversal-ii-Solution1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> LevelOrderBottom(TreeNode root) {
        var result = new List<IList<int>>();
        if(root == null) return result;

        var currLevel = new List<TreeNode> {root};

        while(currLevel.Any()){
            var currLevelVal = new List<int>();
            var newCurrLevel = new List<TreeNode>();
            foreach(var node in currLevel){
                currLevelVal.Add(node.val);
                if(node.left != null) newCurrLevel.Add(node.left);
                if(node.right != null) newCurrLevel.Add(node.right);
            }
            currLevel = newCurrLevel;
            result.Insert(0, currLevelVal);
        }

        return result;
    }
}